datasource/src/test/scala/quasar/physical/mongo/StageInterpreterSpec.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+
+import cats.effect.IO
+import cats.syntax.foldable._
+import cats.instances.list._
+
+import org.bson.{Document => _, _}
+import org.bson.types.Decimal128
+import org.mongodb.scala._
+import org.typelevel.jawn.{AsyncParser, Facade}
+
+import quasar.contrib.std.errorImpossible
+import quasar.{JsonSpec, ScalarStages}
+
+import qdata.QDataEncode
+import qdata.json.QDataFacade
+import qdata.time.{DateTimeInterval, OffsetDate}
+
+import java.time.{
+  LocalDate,
+  LocalDateTime,
+  LocalTime,
+  OffsetDateTime,
+  OffsetTime
+}
+
+import spire.math.Real
+
+import shims._
+
+trait StageInterpreterSpec extends JsonSpec {
+  import MongoSpec.mkMongo
+
+  val E = MongoExpression
+
+  import MongoResource._
+
+  type JsonElement = BsonValue
+
+  def mkCollection(name: String): Collection = Collection(Database("aggregation_test"), name)
+
+  def dropCollection(collection: Collection, mongo: Mongo[IO]): IO[Unit] = {
+    Mongo.singleObservableAsF[IO, Completed](mongo.getCollection(collection).drop())
+      .attempt
+      .map(_ => ())
+  }
+
+  def insertValues(collection: Collection, mongo: Mongo[IO], vals: List[BsonValue]): IO[Unit] = {
+    def docs: List[Document] = vals foldMap {
+      case x: BsonDocument => List(Document(x))
+      case _ => List()
+    }
+    Mongo.singleObservableAsF[IO, Completed](
+      mongo.getCollection(collection) insertMany docs) map (_ => ())
+  }
+
+  val uniqueCollection: IO[Collection] =
+    IO.delay(java.util.UUID.randomUUID().toString) map mkCollection
+
+  implicit def bsonValueQDataEncode: QDataEncode[BsonValue] = new QDataEncode[BsonValue] {
+    def makeLong(l: Long): BsonValue = new BsonInt64(l)
+    def makeDouble(d: Double): BsonValue = new BsonDouble(d)
+    def makeReal(r: Real): BsonValue = {
+      val bd = r.toRational.toBigDecimal(java.math.MathContext.DECIMAL128)
+      val dec128 = new Decimal128(bd.bigDecimal)
+      new BsonDecimal128(dec128)
+    }
+    def makeString(s: String): BsonValue = new BsonString(s)
+    def makeNull: BsonValue = new BsonNull()
+    def makeBoolean(b: Boolean): BsonValue = new BsonBoolean(b)
+    def makeLocalDateTime(l: LocalDateTime) = errorImpossible
```

Comment:
We recently added the rest of the temporal scalars to `ColumnType` in quasar, so they should be able to pass through for `ColumnType.Top` now.

Reply:
Wouldn't this break contract that `QDataEncode` and `QDataDecode` are reversing, since `QDataDecode`, `getLocalDateTime` and all other except `getOffsetDateTime` emit `errorImpossible`

Reply:
Yeah, if that's the scenario for `BSON` the :+1: , but why did you have to redefine it here then?

Reply:
Or did we just not have a `QDataEncode` for `BSON`?

Reply:
Right, we don't have it because why do we need it except tests? :)